src/core/effects/qgseffectpainter.cpp

```diff
@@ -0,0 +1,31 @@
+/***************************************************************************
+                            qgseffectpainter.cpp
+                            ------------------
+    begin                : November 2016
+    copyright            : (C) 2016 Mathieu Pellerin
+    email                : nirvn dot asia at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgseffectpainter.h"
+
+#include <QPainter>
+
+QgsEffectPainter::QgsEffectPainter( QPainter* painter )
+    : mPainter( painter )
+{
+  mPainter->save();
```

Comment:
I think @m-kuhn was also thinking of moving the

    effect->begin( context.renderContext() );

and 

    effect->end( context.renderContext() );

inside here. You'd also then need to store the renderContext and effect here. I'd also suggest moving this to qgspainteffect.h/qgspainteffect.cpp since it's such a small class.